File: PopulateNextPointers.js
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}

    public Node(int _val,Node _left,Node _right,Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

//Time complexity - o(n)
//Space complexity - o(1)
//Brute force -> level order usign queue extra space
class Solution {
    public Node connect(Node root) {
        Node start = root;
        while(start!=null){
            Node current= start;
            while(current!=null){
                if(current.left!=null){
                    current.left.next=current.right;
                }
                if(current.right!=null && current.next!=null){
                    current.right.next=current.next.left;
                }
                current=current.next;
            }
            start=start.left;
        }
       return root; 
        
    }
}

File: PopulateNextPointers.py
#time complexity: O(n)
#Space complexity: O(h)
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None
        self.helper(root.left,root.right)
        return root
    def helper(self,left,right):
        if not left:
            return
        left.next=right
        self.helper(left.left,left.right)
        self.helper(left.right,right.left)
        self.helper(right.left,right.right)
        
        